File: Archivos/APE-5/Ape5Ejercicio9.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;

// ============================================================================
// CLASE VENTA (Aplicación estricta de POO)
// ============================================================================
class Venta {
    // Atributos privados (Encapsulamiento)
    private String producto;
    private int cantidad;
    private double precio;
    private double total;

    // Constructor por defecto
    public Venta() {
        this.producto = "";
        this.cantidad = 0;
        this.precio = 0.0;
        this.total = 0.0;
    }

    // Métodos Setters y Getters
    public void setProducto(String producto) { this.producto = producto; }
    public void setCantidad(int cantidad) { this.cantidad = cantidad; }
    public void setPrecio(double precio) { this.precio = precio; }

    public String getProducto() { return producto; }
    public int getCantidad() { return cantidad; }
    public double getPrecio() { return precio; }
    public double getTotal() { return total; }

    // Método para calcular el total de forma autónoma
    public void calcularTotal() {
        this.total = this.cantidad * this.precio;
    }

    // REQUERIMIENTO: Guardar el registro actual en el archivo de texto
    public void guardarEnArchivo(String nombreArchivo) {
        // Habilitamos el modo append (true) para añadir filas al final del archivo
        try (FileWriter fw = new FileWriter(nombreArchivo, true);
             PrintWriter archivoEscritura = new PrintWriter(fw)) {
            
            // Guardamos los atributos separados por el delimitador '|'
            archivoEscritura.println(producto + "|" + cantidad + "|" + precio + "|" + total);
            
        } catch (IOException e) {
            System.out.println("  [Error] No se pudo escribir en el archivo: " + e.getMessage());
        }
    }

    // REQUERIMIENTO: Mostrar la factura en pantalla
    public void generarFactura() {
        System.out.println("\n=========================================");
        System.out.println("           FACTURA DE VENTA              ");
        System.out.println("=========================================");
        System.out.println("  Producto:   " + producto);
        System.out.println("  Cantidad:   " + cantidad);
        System.out.printf("  Precio Unit: $%.2f\n", precio);
        System.out.println("  ---------------------------------------");
        System.out.printf("  TOTAL:      $%.2f\n", total);
        System.out.println("=========================================");
    }
}

// ============================================================================
// CLASE PRINCIPAL: SistemaVentas
// ============================================================================
public class Ape5Ejercicio9 {

    private static final String NOMBRE_ARCHIVO = "ventas.txt";
    private static final Scanner teclado = new Scanner(System.in);

    // REQUERIMIENTO: Mostrar historial leyendo desde el archivo
    public static void mostrarHistorialVentas() {
        File archivo = new File(NOMBRE_ARCHIVO);

        if (!archivo.exists()) {
            System.out.println("\n[Aviso] No existen ventas registradas en el historial.");
            return;
        }

        double granTotalAcumulado = 0.0;
        int contadorVentas = 0;

        System.out.println("\n====================================================================");
        System.out.println("                     HISTORIAL GENERAL DE VENTAS                    ");
        System.out.println("====================================================================");
        System.out.printf("  %-20s %-12s %-15s %-15s\n", "Producto", "Cantidad", "Precio Unit.", "Total Venta");
        System.out.println("  ------------------------------------------------------------------");

        try (FileReader fr = new FileReader(archivo);
             BufferedReader archivoLectura = new BufferedReader(fr)) {

            String linea;
            while ((linea = archivoLectura.readLine()) != null) {
                if (linea.trim().isEmpty()) continue;

                // Segmentamos la línea usando la barra vertical como expresión regular
                String[] partes = linea.split("\\|");

                if (partes.length == 4) {
                    String prod = partes[0];
                    int cant = Integer.parseInt(partes[1]);
                    double prec = Double.parseDouble(partes[2]);
                    double tot = Double.parseDouble(partes[3]);

                    granTotalAcumulado += tot;
                    contadorVentas++;

                    // Mostrar fila con formato ordenado
                    System.out.printf("  %-20s %-12d $%-14.2f $%-14.2f\n", prod, cant, prec, tot);
                }
            }

            System.out.println("====================================================================");
            System.out.println("  Total de ventas procesadas: " + contadorVentas);
            System.out.printf("  INGRESOS TOTALES ACUMULADOS: $%.2f\n", granTotalAcumulado);
            System.out.println("====================================================================");

        } catch (IOException | NumberFormatException e) {
            System.out.println("Error al procesar el archivo de historial: " + e.getMessage());
        }
    }

    // PROGRAMA PRINCIPAL
    public static void main(String[] args) {
        int opcion = 0;

        do {
            System.out.println("\n==================================");
            System.out.println("        SISTEMA DE VENTAS        ");
            System.out.println("==================================");
            System.out.println("1. Registrar Nueva Venta (Facturar)");
            System.out.println("2. Mostrar Historial de Ventas");
            System.out.println("3. Salir del programa");
            System.out.println("----------------------------------");
            System.out.print("Seleccione una opcion (1-3): ");

            if (teclado.hasNextInt()) {
                opcion = teclado.nextInt();

                switch (opcion) {
                    case 1:
                        Venta nuevaVenta = new Venta(); // Instanciación del objeto
                        
                        System.out.println("\n--- INGRESO DE DATOS DE LA VENTA ---");
                        System.out.print("  Nombre del producto: ");
                        teclado.nextLine(); // Limpieza del buffer del Scanner
                        String prod = teclado.nextLine();
                        nuevaVenta.setProducto(prod);

                        // Validación de cantidad mayor a cero
                        int cant;
                        while (true) {
                            System.out.print("  Cantidad: ");
                            if (teclado.hasNextInt()) {
                                cant = teclado.nextInt();
                                if (cant > 0) break;
                            } else {
                                teclado.next(); // Descartar caracter erróneo
                            }
                            System.out.println("  [Error] Cantidad debe ser un entero mayor a 0.");
                        }
                        nuevaVenta.setCantidad(cant);

                        // Validación de precio mayor a cero
                        double prec;
                        while (true) {
                            System.out.print("  Precio unitario: ");
                            if (teclado.hasNextDouble()) {
                                prec = teclado.nextDouble();
                                if (prec > 0.0) break;
                            } else {
                                teclado.next(); // Descartar caracter erróneo
                            }
                            System.out.println("  [Error] El precio debe ser un numero mayor a 0.");
                        }
                        nuevaVenta.setPrecio(prec);

                        // Ejecución de la lógica interna mediante métodos del objeto
                        nuevaVenta.calcularTotal();
                        nuevaVenta.guardarEnArchivo(NOMBRE_ARCHIVO);
                        nuevaVenta.generarFactura();
                        break;

                    case 2:
                        mostrarHistorialVentas();
                        break;

                    case 3:
                        System.out.println("\nSaliendo del sistema de ventas... ¡Hasta luego!");
                        break;

                    default:
                        System.out.println("[Error] La opcion seleccionada no existe.");
                }
            } else {
                System.out.println("[Error] Entrada no valida. Debe ingresar un numero.");
                teclado.next(); // Limpieza del token inválido
                opcion = 0;
            }

        } while (opcion != 3);

        teclado.close();
    }
}

File: Archivos/APE-5/Ape5Ejercicio9.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <iomanip>
#include <sstream>

using namespace std;

// ============================================================================
// CLASE VENTA (Aplicación de POO)
// ============================================================================
class Venta {
private:
    // Atributos encapsulados
    string producto;
    int cantidad;
    double precio;
    double total;

public:
    // Constructor por defecto
    Venta() {
        producto = "";
        cantidad = 0;
        precio = 0.0;
        total = 0.0;
    }

    // Métodos Setters y Getters (Métodos de acceso)
    void setProducto(string prod) { producto = prod; }
    void setCantidad(int cant) { cantidad = cant; }
    void setPrecio(double prec) { precio = prec; }
    
    string getProducto() { return producto; }
    int getCantidad() { return cantidad; }
    double getPrecio() { return precio; }
    double getTotal() { return total; }

    // Método para calcular el total automáticamente
    void calcularTotal() {
        total = cantidad * precio;
    }

    // REQUERIMIENTO: Guardar el registro en el archivo de texto
    void guardarEnArchivo() {
        // Modo ios::app para añadir datos al final sin borrar lo existente
        ofstream archivoEscritura("ventas.txt", ios::app);
        
        if (!archivoEscritura.is_open()) {
            cout << "  [Error] No se pudo abrir el archivo para registrar la venta.\n";
            return;
        }

        // Guardamos los atributos separados por el delimitador '|'
        archivoEscritura << producto << "|" << cantidad << "|" << precio << "|" << total << "\n";
        archivoEscritura.close();
    }

    // REQUERIMIENTO: Mostrar la factura de la venta actual en pantalla
    void generarFactura() {
        cout << "\n=========================================\n";
        cout << "           FACTURA DE VENTA              \n";
        cout << "=========================================\n";
        cout << "  Producto:   " << producto << "\n";
        cout << "  Cantidad:   " << cantidad << "\n";
        cout << "  Precio Unit:" << fixed << setprecision(2) << " $" << precio << "\n";
        cout << "  ---------------------------------------\n";
        cout << "  TOTAL:      " << fixed << setprecision(2) << " $" << total << "\n";
        cout << "=========================================\n";
    }
};

// ============================================================================
// FUNCIÓN GLOBAL: HISTORIAL DE VENTAS (Manejo de Archivos)
// ============================================================================
void mostrarHistorialVentas() {
    ifstream archivoLectura("ventas.txt");
    
    if (!archivoLectura.is_open()) {
        cout << "\n[Aviso] No existen ventas registradas en el historial (Archivo no encontrado).\n";
        return;
    }

    string linea;
    double granTotalAcumulado = 0.0;
    int contadorVentas = 0;

    cout << "\n====================================================================\n";
    cout << "                     HISTORIAL GENERAL DE VENTAS                    \n";
    cout << "====================================================================\n";
    cout << "  " << left << setw(20) << "Producto" 
         << setw(12) << "Cantidad" 
         << setw(15) << "Precio Unit." 
         << setw(15) << "Total Venta" << "\n";
    cout << "  ------------------------------------------------------------------\n";

    // Ciclo para leer el archivo línea por línea
    while (getline(archivoLectura, linea)) {
        if (linea.empty()) continue;

        stringstream ss(linea);
        string prod, cantStr, precStr, totStr;

        // Descomposición de la línea usando el delimitador '|'
        getline(ss, prod, '|');
        getline(ss, cantStr, '|');
        getline(ss, precStr, '|');
        getline(ss, totStr, '|');

        // Conversión de las cadenas leídas a sus tipos de datos correspondientes
        int cant = stoi(cantStr);
        double prec = stod(precStr);
        double tot = stod(totStr);

        granTotalAcumulado += tot;
        contadorVentas++;

        // Imprimir la fila con formato alineado
        cout << "  " << left << setw(20) << prod 
             << setw(12) << cant 
             << " $" << setw(13) << fixed << setprecision(2) << prec 
             << " $" << setw(13) << fixed << setprecision(2) << tot << "\n";
    }

    archivoLectura.close();

    cout << "====================================================================\n";
    cout << "  Total de ventas procesadas: " << contadorVentas << "\n";
    cout << "  INGRESOS TOTALES ACUMULADOS: $" << fixed << setprecision(2) << granTotalAcumulado << "\n";
    cout << "====================================================================\n";
}

// ============================================================================
// PROGRAMA PRINCIPAL
// ============================================================================
int main() {
    int opcion;
    
    do {
        cout << "\n==================================\n";
        cout << "        SISTEMA DE VENTAS        \n";
        cout << "==================================\n";
        cout << "1. Registrar Nueva Venta (Facturar)\n";
        cout << "2. Mostrar Historial de Ventas\n";
        cout << "3. Salir del programa\n";
        cout << "----------------------------------\n";
        cout << "Seleccione una opcion (1-3): ";
        
        // Validación para evitar bucles infinitos con caracteres inválidos
        if (!(cin >> opcion)) {
            cout << "[Error] Entrada invalida. Intente de nuevo.\n";
            cin.clear();
            cin.ignore(10000, '\n');
            continue;
        }

        switch (opcion) {
            case 1: {
                Venta nuevaVenta; // Instanciación del objeto
                string prod;
                int cant;
                double prec;

                cout << "\n--- INGRESO DE DATOS DE LA VENTA ---\n";
                cout << "  Nombre del producto: ";
                cin.ignore(); // Limpia el buffer
                getline(cin, prod);
                nuevaVenta.setProducto(prod);

                // Validación de cantidad
                do {
                    cout << "  Cantidad: ";
                    if (cin >> cant && cant > 0) break;
                    cout << "  [Error] Cantidad debe ser un entero mayor a 0.\n";
                    cin.clear(); cin.ignore(10000, '\n');
                } while (true);
                nuevaVenta.setCantidad(cant);

                // Validacion de precio
                do {
                    cout << "  Precio unitario: ";
                    if (cin >> prec && prec > 0.0) break;
                    cout << "  [Error] El precio debe ser un numero mayor a 0.\n";
                    cin.clear(); cin.ignore(10000, '\n');
                } while (true);
                nuevaVenta.setPrecio(prec);

                // Procesamiento mediante métodos del objeto
                nuevaVenta.calcularTotal();
                nuevaVenta.guardarEnArchivo();
                nuevaVenta.generarFactura();
                break;
            }
            case 2:
                mostrarHistorialVentas();
                break;
            case 3:
                cout << "\nSaliendo del sistema de ventas... ¡Hasta luego!\n";
                break;
            default:
                cout << "[Error] La opcion seleccionada no existe.\n";
        }
    } while (opcion != 3);

    return 0;
}



